design/src/rfcs/rfc0012_body_callback_apis.md

```diff
@@ -0,0 +1,259 @@
+                    .for_each(|callback| callback.update(bytes));
+            }
+            // When we're done polling for bytes, run each callback's `finally()` method. If any calls to
+            // `finally()` return an error, propagate that error up. Otherwise, continue.
+            Poll::Ready(None) => {
+                for callback_result in this.read_callbacks.iter().map(ReadCallback::finally) {
+                    if let Err(e) = callback_result {
+                        return Poll::Ready(Some(Err(e)));
+                    }
+                }
+            }
+            _ => (),
+        }
+
+        // Now that we've inspected the polling result, all that's left to do is to return it.
+        polling_result
+    }
+
+    // This function now has the added responsibility of cloning callback functions (but with fresh state)
+    // in the case that the `SdkBody` needs to be rebuilt.
+    pub fn try_clone(&self) -> Option<Self> {
+        self.rebuild.as_ref().map(|rebuild| {
+            let next = rebuild();
+            let read_callbacks = self
+                .read_callbacks
+                .iter()
+                .map(ReadCallback::make_new)
+                .collect();
+
+            Self {
+                inner: next,
+                rebuild: self.rebuild.clone(),
+                read_callbacks,
+            }
+        })
+    }
+
+    pub fn with_read_callback(&mut self, read_callback: Box<dyn ReadCallback>) -> &mut Self {
+        self.read_callbacks.push(read_callback);
+        self
+    }
+}
+
+impl http_body::Body for SdkBody {
+    // The other methods have been omitted because they haven't changed
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap<HeaderValue>>, Self::Error>> {
+        let mut last_header_key_seen = None;
+        let header_map = self
+            .read_callbacks
+            .iter()
+            .filter_map(|callback| callback.trailers())
+            .reduce(|mut left_header_map, mut right_header_map| {
+                right_header_map.into_iter().for_each(|(key, value)| {
+                    // For each yielded item that has None provided for the `HeaderName`,
+                    // then the associated header name is the same as that of the previously
+                    // yielded item. The first yielded item will have `HeaderName` set.
+                    // https://docs.rs/http/latest/http/header/struct.HeaderMap.html#method.into_iter-2
+                    match (last_header_key_seen, key) {
+                        (_, Some(key)) => {
+                            left_header_map.append(key, value);
+                            last_header_key_seen = Some(key);
+                        }
+                        (Some(key), None) => {
+                            left_header_map.append(key, value);
+                        }
+                        (None, None) => unreachable!(),
```

Comment:
I think this is one of those cases where a basic `for` loop is probably easier to follow.